MdeModulePkg/Core/Dxe/Mem/Page.c

```diff
@@ -1183,6 +1183,13 @@ CoreFindFreePagesI (
       continue;
     }
 
+    //
+    // Don't allocate out of Special-Purpose memory.
+    //
+    if ((Entry->Attribute & EFI_MEMORY_SP) != 0) {

```

Comment:
I agree with this, however, I think we should add a memory promotion here, where if DXE fails to allocate non-SP memory, that it will then allocate the SP memory. We don't want to fail to boot because we are out of memory but really have some memory we could use. The OS would rather use 90% of the SP memory than nothing because we didn't boot. We can also imagine systems that have very limited (or no) local memory and only CXL attached memory, in which case we need to allocate the SP memory.

Reply:
Per the mail thread, I agree with going with the minimal case for now, while we don't really have systems with EFI_MEMORY_SP in the wild today. As these systems get fleshed out, we can revisit and build an architecture that makes sense then, including the memory promotion scheme I mentioned.